File: app/src/main/java/com/drdisagree/iconify/xposed/utils/SystemUtil.java
package com.drdisagree.iconify.xposed.utils;

/*
 * From AOSPMods
 * https://github.com/siavash79/AOSPMods/blob/canary/app/src/main/java/sh/siava/AOSPMods/utils/SystemUtils.java
 */

import android.annotation.SuppressLint;
import android.app.Application;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.os.Build;
import android.os.Process;
import android.util.TypedValue;

import androidx.annotation.AttrRes;

import com.topjohnwu.superuser.Shell;

public class SystemUtil {

    @SuppressLint("StaticFieldLeak")
    static SystemUtil instance;
    static boolean darkSwitching = false;
    Context mContext;

    public SystemUtil(Context context) {
        mContext = context;
        instance = this;
    }

    public static boolean isDarkMode() {
        if (instance == null) return false;
        return instance.getIsDark();
    }

    public static void doubleToggleDarkMode() {
        boolean isDark = isDarkMode();
        new Thread(() -> {
            try {
                while (darkSwitching) {
                    Thread.currentThread().wait(100);
                }
                darkSwitching = true;

                Shell.cmd("cmd uimode night " + (isDark ? "no" : "yes")).exec();
                Thread.sleep(1000);
                Shell.cmd("cmd uimode night " + (isDark ? "yes" : "no")).exec();
                Thread.sleep(500);

                darkSwitching = false;
            } catch (Exception ignored) {
            }
        }).start();
    }

    public static int getColorResCompat(Context context, @AttrRes int id) {
        TypedValue typedValue = new TypedValue();
        Resources.Theme theme = context.getTheme();
        theme.resolveAttribute(id, typedValue, false);
        @SuppressLint("Recycle") TypedArray arr = context.obtainStyledAttributes(typedValue.data, new int[]{id});
        return arr.getColor(0, -1);
    }

    private boolean getIsDark() {
        return (mContext.getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_YES) == Configuration.UI_MODE_NIGHT_YES;
    }

    public static <Method> void killSelf() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            BootLoopProtector.resetCounter(Process.myProcessName());
        } else {
            BootLoopProtector.resetCounter(Application.getProcessName());
        }
        Process.killProcess(Process.myPid());
    }
}

File: app/src/main/java/com/drdisagree/iconify/xposed/utils/SystemUtil.kt
package com.drdisagree.iconify.xposed.utils

import android.annotation.SuppressLint
import android.app.Application
import android.content.Context
import android.content.res.Configuration
import android.os.Build
import android.os.Process
import com.drdisagree.iconify.xposed.utils.BootLoopProtector.resetCounter
import com.topjohnwu.superuser.Shell
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

class SystemUtil(var mContext: Context) {

    init {
        instance = this
    }

    private val isDark: Boolean
        get() = mContext.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_YES == Configuration.UI_MODE_NIGHT_YES

    companion object {
        @SuppressLint("StaticFieldLeak")
        var instance: SystemUtil? = null
        private var darkSwitching = false
        val isDarkMode: Boolean
            get() = if (instance == null) false else instance!!.isDark

        fun doubleToggleDarkMode() {
            val isDark = isDarkMode

            CoroutineScope(Dispatchers.Default).launch {
                try {
                    while (darkSwitching) {
                        delay(100)
                    }

                    darkSwitching = true

                    Shell.cmd("cmd uimode night ${if (isDark) "no" else "yes"}").exec()
                    delay(1000)
                    Shell.cmd("cmd uimode night ${if (isDark) "yes" else "no"}").exec()
                    delay(500)

                    darkSwitching = false
                } catch (ignored: Exception) {
                }
            }
        }

        fun <Method> killSelf() {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                resetCounter(Process.myProcessName())
            } else {
                resetCounter(Application.getProcessName())
            }
            Process.killProcess(Process.myPid())
        }
    }
}
